### PR TITLE
Prune workflows based on changed files

### PR DESCRIPTION
**Pull request by KyleFromNVIDIA**
Only run tests based on things that have actually changed. For example, if only Python files have changed, we don't need to run the C++ tests.

Contributes to https://github.com/rapidsai/build-planning/issues/94

Depends on https://github.com/rapidsai/shared-workflows/pull/239

### CODE REVIEW

File: .github/workflows/pr.yaml
Comment (robertmaynard) [inline .github/workflows/pr.yaml:68]: I think we can add the following folders for cpp:

- mg_utils
- readme_pages

File: .github/workflows/pr.yaml
Comment (KyleFromNVIDIA) [inline .github/workflows/pr.yaml:167]: This change results in the `cugraph` wheel not being built if Python tests aren't needed, because `wheel-build-cugraph` depends on `wheel-tests-pylibcugraph`. Is it OK to skip building the `cugraph` wheel in that case?

File: notebooks/algorithms/structure/Renumber.ipynb
Comment (acostadon) [inline notebooks/algorithms/structure/Renumber.ipynb:64]: Should this have been removed?

Comment (KyleFromNVIDIA) [reply notebooks/algorithms/structure/Renumber.ipynb:64]: I put it in for testing. The thing I tested was successful, so I'll remove it.

File: .github/workflows/pr.yaml
Comment (jameslamb) [inline .github/workflows/pr.yaml:46]: I think `!.devcontainers/**` could also safely be added to these lists. A change that just updates the base image for devcontainers shouldn't require running any of the other tests in this repo.

File: .github/workflows/pr.yaml
Comment (jameslamb) [reply .github/workflows/pr.yaml]: > Is it OK to skip building the cugraph wheel in that case?

I think it is ok, yes.

But only if I've understood how this works correctly.

Seeing these paths in the `test_python` config:

```text
CONTRIBUTING.md
README.md
docs/**
img/**
notebooks/
```

I'm assuming that translates to "if any file was changed that doesn't match those patterns, then run the builds and test".

I cannot think of a case where a PR touching only those files would require rebuilding wheels...given that the notebooks are tested with conda packages, not wheels.

**Review by jameslamb — COMMENTED** [review]
I'm not up to date on the changed-files work, so my review shouldn't count towards a merge. I left one other comment for your consideration.

I'll get caught up today and could leave a more thoughtful review later today, if you still don't have a `ci-codeowners` approval by then.

File: .github/workflows/pr.yaml
Comment (vyasr) [reply .github/workflows/pr.yaml]: Yes, I think that is OK, although it is a bit incongruous to build pylibcugraph but not cugraph. We could also reorder the jobs to make `wheel-build-cugraph` depend on `wheel-build-pylibcugraph` instead.

File: .github/workflows/pr.yaml
Comment (KyleFromNVIDIA) [reply .github/workflows/pr.yaml:46]: I've added `!.devcontainers/**`.

File: .github/workflows/pr.yaml
Comment (KyleFromNVIDIA) [reply .github/workflows/pr.yaml]: If nobody has any objections, I'll merge this as-is, and we can consider revisiting the build graph at some point in the future if it becomes a problem.

**Review by robertmaynard — APPROVED** [review]
C++ build side rules look good

**Review by AyodeAwe — APPROVED** [review]

**Pull request merged**
